src/common/procedure/src/store/state_store.rs

```diff
@@ -64,49 +65,49 @@ impl ObjectStateStore {
 #[async_trait]
 impl StateStore for ObjectStateStore {
     async fn put(&self, key: &str, value: Vec<u8>) -> Result<()> {
-        let object = self.store.object(key);
-        object.write(value).await.context(PutStateSnafu { key })
+        self.store
+            .write(key, value)
+            .await
+            .context(PutStateSnafu { key })
     }
 
     async fn walk_top_down(&self, path: &str) -> Result<KeyValueStream> {
         let path_string = path.to_string();
 
-        let lister = self
-            .store
-            .object(path)
-            .scan()
-            .await
-            .map_err(|e| Error::ListState {
-                path: path_string.clone(),
-                source: e,
-            })?;
+        let mut lister = self.store.scan(path).await.map_err(|e| Error::ListState {
+            path: path_string.clone(),
+            source: e,
+        })?;
 
-        let stream = lister
-            .try_filter_map(|entry| async move {
+        let store = self.store.clone();
+
+        let stream = try_stream!({
+            while let Some(res) = lister.next().await {
+                let entry = res.context(ListStateSnafu { path: &path_string })?;
                 let key = entry.path();
-                let key_value = match entry.mode().await? {
-                    ObjectMode::FILE => {
-                        let value = entry.read().await?;
-
-                        Some((key.to_string(), value))
-                    }
-                    ObjectMode::DIR | ObjectMode::Unknown => None,
-                };
-
-                Ok(key_value)
-            })
-            .map_err(move |e| Error::ListState {
-                path: path_string.clone(),
-                source: e,
-            });
+                let metadata = store
+                    .metadata(&entry, Metakey::Mode)
+                    .await
+                    .context(ListStateSnafu { path: key })?;
+                if let EntryMode::FILE = metadata.mode() {
+                    let value = store
+                        .read(key)
+                        .await
+                        .context(ListStateSnafu { path: key })?;
+                    yield (key.to_string(), value);
```

Comment:
In the previous code, if it's not a file, it yields `None`. But we change the semantic here. Is it all right?

Reply:
> In the previous code, if it's not a file, it yields `None`. But we change the semantic here. Is it all right?

No, previous code uses a `try_filter_map` that filters out `None`s, and `KeyValueStream` always yields a `Result<KeyValue>` rather than `Result<Option<KeyValue>>`.